### PR TITLE
gh-111085 Fix logic order causing nicer error to never be raised in asyncio.Timeout

### PR DESCRIPTION
**Pull request by Gobot1234**
~~I don't think this needs a news entry or an issue~~

* Issue: gh-111085

### CODE REVIEW

**Review by hauntsaninja — APPROVED** [review]
Thanks for fixing this! However, it does need a NEWS entry since exception type change is user visible behaviour

**Comment by Gobot1234** [comment]
@hauntsaninja does this also need an issue then?

**Comment by AlexWaygood** [comment]
> @hauntsaninja does this also need an issue then?

Yeah, it's not possible to add a news entry without linking the news entry to a GitHub issue

**Review by Eclips4 — APPROVED** [review]
LGTM. Should it be backported to 3.12?

**Comment by Gobot1234** [comment]
Turns out the bug exists on 3.11 as well

File: Misc/NEWS.d/next/Library/2023-10-19-11-40-38.gh-issue-111085.hBPpQr.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix :class:`asyncio.Timeout` raising an :exc:`AttributeError` over a :exc:`RuntimeError` due to logic bug. Patch by James Hilton-Balfe
```
Comment (Eclips4) [inline Misc/NEWS.d/next/Library/2023-10-19-11-40-38.gh-issue-111085.hBPpQr.rst:1]: Just FYI: you can also add yourself to `Misc/ACKS` (please, follow the alphabetical order). I couldn't find you there ;)

**Comment by hauntsaninja** [comment]
I think this probably should not be backported. This effectively just changes a bad error message to a good one and has a chance of breaking code, so doesn't feel like good cost benefit.

**Comment by Gobot1234** [comment]
Closing in favour of #111111

**Comment by serhiy-storchaka** [comment]
While this change fixes an AttributeError in `__enter__()`, it left the object in inconsistent state, so other AttributeError can be raised in specific circumferences in other place. There are also other issues in the current code. It makes sense to fix them all at once.

In any case, thank you for your contribution. I included your changes in the other PR and added you as a co-author.

**Pull request closed without merging**